contracts/mocks/extensions/tokenProperties/RMRKTokenPropertiesMock.sol

```diff
@@ -0,0 +1,157 @@
+
+/**
+ * @title RMRKTokenProperties
+ * @author RMRK team
+ * @notice Smart contract of the RMRK Token properties module.
+ */
+contract RMRKTokenPropertiesMock is IRMRKTokenProperties {
+
+    mapping(string => uint256) private _keysToIds;
+    uint256 private _totalProperties;
+
+    mapping(string => uint256) private _stringValuesToIds;
+    mapping(uint256 => mapping(uint256 => string)) private _stringValues;
+
+    mapping(uint256 => mapping(uint256 => address)) private _addressValues;
+    mapping(uint256 => mapping(uint256 => bytes)) private _bytesValues;
+    mapping(uint256 => mapping(uint256 => uint256)) private _uintValues;
+    mapping(uint256 => mapping(uint256 => bool)) private _boolValues;
+
+    /**
+    * @notice Used to retrieve the string type token properties.
+     * @param tokenId The token ID
+     * @param key The key of the property
+     * @return string The value of the string property
+     */
+    function getStringTokenProperty(uint256 tokenId, string memory key) external view returns (string memory) {
+        return _stringValues[tokenId][_keysToIds[key]];
+    }
+
+    /**
+     * @notice Used to retrieve the uint type token properties.
+     * @param tokenId The token ID
+     * @param key The key of the property
+     * @return uint256 The value of the uint property
+     */
+    function getUintTokenProperty(uint256 tokenId, string memory key) external view returns (uint256) {
+        return _uintValues[tokenId][_keysToIds[key]];
+    }
+
+    /**
+     * @notice Used to retrieve the bool type token properties.
+     * @param tokenId The token ID
+     * @param key The key of the property
+     * @return bool The value of the bool property
+     */
+    function getBoolTokenProperty(uint256 tokenId, string memory key) external view returns (bool) {
+        return _boolValues[tokenId][_keysToIds[key]];
+    }
+
+    /**
+     * @notice Used to retrieve the address type token properties.
+     * @param tokenId The token ID
+     * @param key The key of the property
+     * @return address The value of the address property
+     */
+    function getAddressTokenProperty(uint256 tokenId, string memory key) external view returns (address) {
+        return _addressValues[tokenId][_keysToIds[key]];
+    }
+
+    /**
+     * @notice Used to retrieve the bytes type token properties.
+     * @param tokenId The token ID
+     * @param key The key of the property
+     * @return bytes The value of the bytes property
+     */
+    function getBytesTokenProperty(uint256 tokenId, string memory key) external view returns (bytes memory) {
+        return _bytesValues[tokenId][_keysToIds[key]];
+    }
+
+    /**
+     * @notice Used to set a number property.
+     * @param tokenId The token ID
+     * @param key The property key
+     * @param value The property value
+     */
+    function setUintProperty(uint256 tokenId, string memory key, uint256 value) external {
+        _keysToIds[key] = _totalProperties;
```

Comment:
You need to check if that key was already mapped to an Id for 2 reasons:
1. Optimization, same keys should use the same ID.
2. If it existed before you're rewriting it, and the previous value associated to that key won't be found because when reading you will point to a different id. I suggest using something like this on every setter to get the right Id:
```
function _getIdForKey(string memory key) internal returns (uint256) {
        if (_keysToIds[key] == 0) {
            _totalProperties += 1; // Add first so it never uses 0 as ID
            _keysToIds[key] = _totalProperties;
            return _totalProperties;
        }
        else {
            return _keysToIds[key];
        }
    } 
```

On setters you simply get the Id by `_getIdForKey(key)` without worrying if it existed or not.

Reply:
Added it now for all the setters except strings.

Reply:
As strings setter already checks for duplicate keys and values.